File: LeetCode/278_FirstBadVersion.py
# https://leetcode.com/problems/first-bad-version/
# 2021 11.20


# The isBadVersion API is already defined for you.
# @param version, an integer
# @return an integer
# def isBadVersion(version):
#     return

class Solution:
    def firstBadVersion(self, n):
        """
        :type n: int
        :rtype: int
        """
        # v1
        left,right=1,n
        if n==1:
            return 1
        while left<=right:
            pivot=(left+right)//2
            if isBadVersion(pivot):
                lastpivot=pivot
                right=pivot-1
            else:
                left=pivot+1
        return lastpivot

        # # v2
        # left,right=1,n
        # if n==1:
        #     return 1
        # while left<=right:
        #     pivot=(left+right)//2
        #     if isBadVersion(pivot):
        #         # lastpivot=pivot
        #         right=pivot-1
        #         if not isBadVersion(right):
        #             return pivot
        #     else:
        #         left=pivot+1
        # return pivot

File: LeetCode/278_FirstBadVersion.go
// https://leetcode.com/problems/first-bad-version/
// 2022年07月02日 10:23:55

/** 
 * Forward declaration of isBadVersion API.
 * @param   version   your guess about first bad version
 * @return 	 	      true if current version is bad 
 *			          false if current version is good
 * func isBadVersion(version int) bool;
 */

 func firstBadVersion(n int) int {
    l,h:=0,n
    for l<=h {
        mid:=(l+h)/2
        if isBadVersion(mid) {
            h=mid-1
        }else{
            l=mid+1
        }
    }
    return l
}